File: Core-DSA-Optimized/Tress/Treap/treap.cpp
#include <iostream>
#include <cstdlib>
#include <cassert>

using namespace std;

struct TreapNode {
    int key;
    double priority;
    TreapNode* left;
    TreapNode* right;

    TreapNode(int k) : key(k), left(nullptr), right(nullptr) {
        priority = static_cast<double>(rand()) / RAND_MAX;
    }
};

class Treap {
private:
    TreapNode* rightRotate(TreapNode* y) {
        TreapNode* x = y->left;
        TreapNode* T2 = x->right;
        x->right = y;
        y->left = T2;
        return x;
    }

    TreapNode* leftRotate(TreapNode* x) {
        TreapNode* y = x->right;
        TreapNode* T2 = y->left;
        y->left = x;
        x->right = T2;
        return y;
    }

    TreapNode* insertNode(TreapNode* node, int key) {
        if (node == nullptr) return new TreapNode(key);

        if (key < node->key) {
            node->left = insertNode(node->left, key);
            if (node->left->priority > node->priority) {
                node = rightRotate(node);
            }
        } else if (key > node->key) {
            node->right = insertNode(node->right, key);
            if (node->right->priority > node->priority) {
                node = leftRotate(node);
            }
        }
        return node;
    }

    TreapNode* deleteNode(TreapNode* node, int key) {
        if (node == nullptr) return nullptr;

        if (key < node->key) {
            node->left = deleteNode(node->left, key);
        } else if (key > node->key) {
            node->right = deleteNode(node->right, key);
        } else {
            if (node->left == nullptr && node->right == nullptr) {
                delete node;
                return nullptr;
            }
            if (node->left == nullptr) {
                TreapNode* temp = node->right;
                delete node;
                return temp;
            }
            if (node->right == nullptr) {
                TreapNode* temp = node->left;
                delete node;
                return temp;
            }

            if (node->left->priority < node->right->priority) {
                node = leftRotate(node);
                node->left = deleteNode(node->left, key);
            } else {
                node = rightRotate(node);
                node->right = deleteNode(node->right, key);
            }
        }
        return node;
    }

    void cleanup(TreapNode* node) {
        if (node != nullptr) {
            cleanup(node->left);
            cleanup(node->right);
            delete node;
        }
    }

public:
    TreapNode* root;

    Treap() {
        root = nullptr;
        srand(42); // Deterministic testing
    }

    ~Treap() {
        cleanup(root);
    }

    void insert(int key) {
        root = insertNode(root, key);
    }

    void remove(int key) {
        root = deleteNode(root, key);
    }

    bool search(int key) {
        TreapNode* curr = root;
        while (curr != nullptr) {
            if (curr->key == key) return true;
            if (key < curr->key) curr = curr->left;
            else curr = curr->right;
        }
        return false;
    }
};

// --- CI/CD Automated Test ---
int main() {
    Treap treap;
    for (int i = 1; i <= 7; i++) treap.insert(i);

    assert(treap.search(4) == true);
    assert(treap.search(10) == false);
    assert(treap.root->key != 1);

    treap.remove(4);
    assert(treap.search(4) == false);

    cout << "C++ Treap Test Passed! (Root balanced to: " << treap.root->key << ")\n";
    return 0;
}

File: Core-DSA-Optimized/Tress/Treap/treap.py
import random

class TreapNode:
    def __init__(self, key: int):
        self.key = key
        # Random priority determines heap structure
        self.priority = random.random()
        self.left = None
        self.right = None

class Treap:
    def __init__(self):
        self.root = None

    def _right_rotate(self, y: TreapNode) -> TreapNode:
        x = y.left
        T2 = x.right
        x.right = y
        y.left = T2
        return x

    def _left_rotate(self, x: TreapNode) -> TreapNode:
        y = x.right
        T2 = y.left
        y.left = x
        x.right = T2
        return y

    def _insert(self, root: TreapNode, key: int) -> TreapNode:
        if root is None:
            return TreapNode(key)

        # Standard BST Insert
        if key < root.key:
            root.left = self._insert(root.left, key)
            # Fix Heap property if violated
            if root.left.priority > root.priority:
                root = self._right_rotate(root)
        elif key > root.key:
            root.right = self._insert(root.right, key)
            # Fix Heap property if violated
            if root.right.priority > root.priority:
                root = self._left_rotate(root)
        
        return root

    def insert(self, key: int):
        self.root = self._insert(self.root, key)

    def _delete(self, root: TreapNode, key: int) -> TreapNode:
        if root is None:
            return root

        if key < root.key:
            root.left = self._delete(root.left, key)
        elif key > root.key:
            root.right = self._delete(root.right, key)
        else:
            # Node found. If it's a leaf, simply drop it.
            if root.left is None and root.right is None:
                return None
            
            # If one child is empty, return the other
            elif root.left is None:
                return root.right
            elif root.right is None:
                return root.left
            
            # If both children exist, rotate the child with higher priority UP
            elif root.left.priority < root.right.priority:
                root = self._left_rotate(root)
                root.left = self._delete(root.left, key)
            else:
                root = self._right_rotate(root)
                root.right = self._delete(root.right, key)

        return root

    def delete(self, key: int):
        self.root = self._delete(self.root, key)

    def search(self, key: int) -> bool:
        curr = self.root
        while curr:
            if curr.key == key:
                return True
            if key < curr.key:
                curr = curr.left
            else:
                curr = curr.right
        return False

# --- CI/CD Automated Test ---
if __name__ == '__main__':
    # Fix the random seed to ensure deterministic CI/CD testing
    random.seed(42)
    
    treap = Treap()
    
    # Inserting sequential data. A normal BST would degrade to a straight line.
    for i in range(1, 8):
        treap.insert(i)
        
    assert treap.search(4) == True, "Failed to find inserted node!"
    assert treap.search(10) == False, "Found hallucinated node!"
    
    # Because of random priorities, the root should not be 1 (which would indicate a linked list)
    assert treap.root.key != 1, "Treap degraded into a linked list!"
    
    treap.delete(4)
    assert treap.search(4) == False, "Failed to delete node!"
    
    print(f"Python Treap Test Passed! (Root naturally balanced to: {treap.root.key})")